Upgrade Prefect to 3.1.13

dgarros opened the pull request:
This PR is trying to upgrade Prefect to 3.1.13 because we need it for the event management epic as part of 1.2

Prefect made a lot of progress around Typing in 3.1.x but there are still a few issues reported by Mypy. 
I opened some issues on prefect github project to track them and I've disabled mypy for these issues for now 
- https://github.com/PrefectHQ/prefect/issues/16769
- https://github.com/PrefectHQ/prefect/issues/16770

dgarros committed:
Upgrade Prefect to 3.1.13

fatih-acar commented:
Need to investigate slowness on Prefect startup, it actually takes more than 30 seconds to start Prefect when running with xdist (probably because the CI runner is overwhelmed)

Maybe related to https://github.com/PrefectHQ/prefect/issues/16299

tomukmatthews commented:
> Need to investigate slowness on Prefect startup, it actually takes more than 30 seconds to start Prefect when running with xdist (probably because the CI runner is overwhelmed)
> 
> Maybe related to [PrefectHQ/prefect#16299](https://github.com/PrefectHQ/prefect/issues/16299)

I have the same problem. I would be cautious about the latest versions of Prefect, many of our data pipelines are experiencing significant flakiness and degredation at the moment due to the performance related issues of the event driven architecture (internally) of prefect since 3.x. Have you experienced anything like that outside of tests?

fatih-acar committed:
fix(backend, tests): increase sleep time during rmq test

In some cases, when the CI runner is overwhelmed, 0.1s may not be enough
to successfully send a message to rabbitmq. Increase it to 1 second to
be consistent with other tests.

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar committed:
fix(backend, tests): increase prefect startup timeout

30 second may not be enough when using Prefect 3.1 which seems to add a
little overhead to 3.0. Prefect usually took 20s-25s to start and a
small load spike on the CI runner can fail the tests.

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar commented:
> > Need to investigate slowness on Prefect startup, it actually takes more than 30 seconds to start Prefect when running with xdist (probably because the CI runner is overwhelmed)
> > Maybe related to [PrefectHQ/prefect#16299](https://github.com/PrefectHQ/prefect/issues/16299)
> 
> I have the same problem. I would be cautious about the latest versions of Prefect, many of our data pipelines are experiencing significant flakiness and degredation at the moment due to the performance related issues of the event driven architecture (internally) of prefect since 3.x. Have you experienced anything like that outside of tests?

We started using Prefect since 3.0, so we can't tell for previous releases.
But there seems to be a small performance regression between 3.0 and 3.1 that makes our CI pipelines fail due to load on our runners that causes slow startup.
We observe startups taking between 20s to 30s, and sometimes reaching the default timeout value which is 30s.
Fixed that by increasing `server_startup_timeout` in `prefect_test_harness` to 60s.

fatih-acar committed:
feat(backend): enable tracing for task workers

This was previously enabled on git agents but got removed during the
transition to task workers.
Prefect 3.1 also adds instrumentation to workflow runs.

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar committed:
fix(ci): use old Prefect version when testing demo

This is a temporary fix until we use the Infrahub image to run prefect
server.

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar commented:
Hardcoded versions in CI is not the best way to fix this.
I will work on using our own image to run prefect-server in another PR.

fatih-acar commented:
Still need to check the cache key error showing in the server logs
And fix the logger configuration

dgarros committed:
Set cache_policy for setup_triggers task

dgarros commented:
> Still need to check the cache key error showing in the server logs And fix the logger configuration

I pushed a fix for the `cache key` error, it was related to #5387
Let's merge this PR and we'll fix the logging issue in `develop`
